### PR TITLE
NIFI-4786 - Allow Expression Evaluation to Kinesis/Firehose Stream Name

### PR DESCRIPTION
**Pull request by SunSatION**
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by jvwing** [comment]
Reviewing...

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/firehose/PutKinesisFirehose.java [hunk 1]
```diff
@@ -89,63 +89,80 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
...
             for (int i = 0; i < flowFiles.size(); i++) {
                 FlowFile flowFile = flowFiles.get(i);
 
+                final String firehoseStreamName = context.getProperty(KINESIS_FIREHOSE_DELIVERY_STREAM_NAME).evaluateAttributeExpressions(flowFile).getValue();
+
                 final ByteArrayOutputStream baos = new ByteArrayOutputStream();
                 session.exportTo(flowFile, baos);
-                records.add(new Record().withData(ByteBuffer.wrap(baos.toByteArray())));
+
+                if ( !recordHash.containsKey(firehoseStreamName) )
```
Comment (jvwing) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/firehose/PutKinesisFirehose.java]: Thank you for following the pre-existing code style in this class file.  That's usually a good practice.  However, `if` statements in NiFi code typically have braces even for single-line contents, and there is typically no space between parenthesis:
```
if (somevar == true) {
    doStuff();
}
```
Would you please fix these up?

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/PutKinesisStream.java [hunk 2]
```diff
@@ -111,64 +112,80 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
...
+                        .evaluateAttributeExpressions(flowFiles.get(i)).getValue();
 
                 if ( ! StringUtils.isBlank(partitionKey) ) {
                     record.setPartitionKey(partitionKey);
                 } else {
                     record.setPartitionKey(Integer.toString(randomParitionKeyGenerator.nextInt()));
                 }
 
-                records.add(record);
+                if ( !recordHash.containsKey(streamName) ) {
```
Comment (jvwing) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/PutKinesisStream.java]: Would you also please fix these `if`s?  I completely understand they were already that way.

**Comment by jvwing** [comment]
@SunSatION , thanks for submitting this PR, it looks like good stuff.

May I ask how you tested the changes?  For PutKinesisFirehose, I was able to run the integration test suite successfully, and I was able to manually build and run a simple flow sending data to Kinesis Firehose.  For PutKinesisStream, the manual flow building worked, but I get exceptions running the integration test suite (ITPutKinesisStream.testIntegrationSuccess) as it calls `putRecords()`:

```
com.amazonaws.SdkClientException: Unable to marshall request to JSON: 
com.fasterxml.jackson.dataformat.cbor.CBORGenerator.getOutputContext()Lcom/fasterxml/jackson/core/json/JsonWriteContext;
```

Did you try the integration tests, and did they work for you?  I don't believe your change introduced this issue.  A quick check of the master branch suggests it was already there, but it complicated testing.

**Comment by SunSatION** [comment]
Hi @jvwing 

Thanks for the PR review. I've amended some changes to follow NIFI's `if`s convention. 

I'll try having a look at the integration tests in the evening. I've performed the manual tests as you did and used `mvn -Pcontrib-check clean install` in the nifi-aws-bundle folder where now I've realised that it didn't perform the integration testing

**Comment by SunSatION** [comment]
Hi @jvwing 

The problem is caused by an unsupported version of Jackson databind defined in the parent NIFI POM (2.9.3) file compared to the used by AWS Java SDK (2.6.6)

I've included the old version with scope set to `test` in the `nifi-aws-bundle` POM file, however I'm not sure whether this is the right approach. Any suggestions?

The tests were executed successfully. 

[Test Results - ITPutKinesisStream.log](https://github.com/apache/nifi/files/1665817/Test.Results.-.ITPutKinesisStream.log)

**Comment by jvwing** [comment]
@SunSatION, this is nice work.  Thanks for the new Kinesis expression language features.  And double thanks for fixing the if statements and the integration tests, I think your Jackson test dependency fix is a good approach.

**Pull request closed without merging**